### PR TITLE
Rule S3717: Track use of 'NotImplementedException'

### LINKED ISSUE #585: Rule S3717: Track use of "NotImplementedException"
[RSPEC-3717](https://jira.sonarsource.com/browse/RSPEC-3717)

`NotImplementedException` is often used to mark methods which must be implemented for the overall functionality to be complete, but which the developer wants to implement later. That's as opposed to the `NotSupportedException` which is thrown by methods which are required by base classes or interfaces, but which are not appropriate to the current class.

### PR DESCRIPTION
**Pull request by Evangelink**
Fix #585

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TrackNotImplementedException.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+                {
+                    var throwStatement = (ThrowStatementSyntax)c.Node;
+                    if (throwStatement.Expression == null)
+                    {
+                        return;
+                    }
+
+                    var typeInfo = c.SemanticModel.GetTypeInfo(throwStatement.Expression);
+
+                    if (typeInfo.Type == null ||
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TrackNotImplementedException.cs]: This if with return can be ommited - Type.Is already covers these cases.

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/TrackNotImplementedException.cs [hunk 2]
```diff
@@ -0,0 +1,13 @@
+using System;
+
+namespace Tests.Diagnostics
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/TrackNotImplementedException.cs]: Let's add more test cases:
1. Create a type inheriting from NotImplementedException. Should we report on that?
2. Create an exception, but do not throw.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TrackNotImplementedException.cs (hunk 1 above)
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TrackNotImplementedException.cs]: My bad!

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/TrackNotImplementedException.cs (hunk 2 above)
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/TrackNotImplementedException.cs]: Will add the tests. I don't think we should handle sub-classes of `NotImplementedException`.

Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/TrackNotImplementedException.cs]: Fine with that (as long as you add tests).

Should we mention that exception in RSPEC text?

**Pull request merged**